pkg/agent/billing.go

```diff
@@ -93,84 +107,140 @@ func (s *billingMetricsState) collect(conf *BillingConfig, store *util.WatchStor
-				StopTime:       slice.endTime,
-				Value:          int(slice.metrics.cpu),
-			})
-		})
+func (h *vmMetricsHistory) appendSlice(timeSlice metricsTimeSlice) {
+	// Try to extend the existing period of continuous usage
+	if h.lastSlice != nil && h.lastSlice.tryMerge(timeSlice) {
+		return
 	}
 
-	s.historical = make(map[billingMetricsKey]vmMetricsHistory)
-	return batch
+	// Something's new. Push previous time slice, start new one:
+	h.finalizeCurrentTimeSlice()
+	h.lastSlice = &timeSlice
 }
 
-func (h vmMetricsHistory) forEachMergedSlice(process func(metricsTimeSlice)) {
-	if len(h.slices) == 0 {
+// finalizeCurrentTimeSlice pushes the current time slice onto h.total
+//
+// This ends up rounding down the total time spent on a given time slice, so it's best to defer
+// calling this function until it's actually needed.
+func (h *vmMetricsHistory) finalizeCurrentTimeSlice() {
+	if h.lastSlice == nil {
 		return
 	}
 
-	last := h.slices[0]
-	for _, slice := range h.slices[1:] {
-		if last.endTime == slice.startTime && last.metrics == slice.metrics {
-			last.endTime = slice.endTime
-		} else {
-			process(last)
-			last = slice
-		}
+	duration := h.lastSlice.endTime.Sub(h.lastSlice.startTime).Nanoseconds()
+	if duration < 0 {
+		panic("negative duration")
+	}
+
+	// note: we could've used Duration.Seconds() above, but that gives us a float. We absolutely do
+	// not want to be doing billing math with floats.
+	//
+	// TODO: This approach here explicitly rounds us down - otherwise rounding is actually quite
+	// tricky, and it's safer to round down than up. In the future, we might internally store things
+	// as milliseconds or something, and round down to the nearest "CPU second" when sending the
+	// batch, in order to be more accurate. There's a variety of options.
+	seconds := duration / time.Second.Nanoseconds()
+	metricsSeconds := vmMetricsSeconds{
+		cpu: uint32(h.lastSlice.metrics.cpu) * uint32(seconds),
+	}
+	h.total.cpu += metricsSeconds.cpu
+
+	h.lastSlice = nil
+}
 
+// tryMerge attempts to merge s and next (assuming that next is after s), returning true only if
+// that merging was successful.
+//
+// Merging may fail if s.endTime != next.startTime or s.metrics != next.metrics.
+func (s *metricsTimeSlice) tryMerge(next metricsTimeSlice) bool {
+	merged := s.endTime == next.startTime && s.metrics == next.metrics
+	if merged {
+		s.endTime = next.endTime
 	}
+	return merged
+}
+
+// clear updates the state to remove any previously un-pushed history
+func (s *billingMetricsState) clear() {
+	s.historical = make(map[billingMetricsKey]vmMetricsHistory)
+}
+
+// makeBatch creates a billing.Batch for the currently stored history, without modfiying any of it
+func (s *billingMetricsState) makeBatch(client *billing.Client) *billing.Batch {
+	batch := client.NewBatch()
+
+	now := time.Now()
+
+	for key, history := range s.historical {
+		batch.AddIncrementalEvent(billing.IncrementalEvent{
+			IdempotencyKey: uuid.NewString(),
```

Comment:
I feel like it's kind of wrong to use a random string (essentially) rather than the name of the object and some stored increment, there could be some weird http socket behavior that putting a random string here would cause us to think that we'd failed but have the request actually go through.

Reply:
See recent changes, should be handled.